fix: student sharing platform blogpost (#620)
ory/web: src/markdown/blog/student-sharing-platform-open-source-auth.mdx

So the login is as follows:

1. Request a login flow.
1. Kratos returns login flow ID.
1. Send login to Ory Kratos, referencing the flow ID.
1. Kratos returns session cookie.
1. Request protected field from GQL api, attaching session cookie.
1. GQL API asks Ory Kratos to verify session cookie.
1. Kratos returns user ID belonging to session cookie.
1. GQL API returns protected field.

## Redirects and Account recovery

Sometimes, Ory Kratos and the user interface need another concept to interact.
Let's say you forgot your password and requested a recovery link via mail. The
problem Ory developers needed to solve here is to send some _token_ from the
recovery link back to Ory Kratos (to verify the email) and at the same time
associate it with a session of the UI in a web browser so that the user can
reset their password in the UI, and all with minimal knowledge of the specific
app architecture.

They solved this by hosting Ory Kratos on the same domain as the app itself and
using _redirects_. For instance our UI is hosted under
`https://huddle.hsg.fs.tum.de/ui` whereas Ory Kratos is hosted under
`https://huddle.hsg.fs.tum.de/.ory/kratos`. The messages exchanged are as

> follows:

Reviewer: not sure what you mean here. I understand the first paragraph, but the conclusion doesn't make sense. We could have sub domains here if we wanted to - as long as we are on the same TLD the cookies will be set correctly. We could even do something like our app is on `https://huddle.hsg...` while Ory Kratos is on `https://kratos.hsg.fs.tum.de`. They don't need to be separated by paths.

An who are _they_? The first paragraph is talking about Ory Developers, while the second is talking about the developers implementing Ory?

Reviewer 2: The same thing for me since Kratos feels fine in multi-subdomain environments.

Author: > The problem Ory developers needed to solve here is to send some _token_ from the
> recovery link back to Ory Kratos (to verify the email) and at the same time
> associate it with a session of the UI in a web browser so that the user can
> reset their password in the UI

So what do you think when we say here it is solved (i would say "by Ory Kratos") by using browser redirects (plus with built-in security) 
and maybe link to the [multitenancy document](https://www.ory.sh/docs/kratos/guides/multi-tenancy-multitenant) @Benehiko ?

1. User clicks recovery link in the format
   `https://huddle.hsg.fs.tum.de/.ory/kratos/self-service/recovery?flow=6b817dhf-4717-48we-a756-cd7902b43ce6&token=QiGjssJcvB08H7cUdE3tyXvQr1i4OSaR`.
   Note that this will result in a GET request by the browser that will be
   handled by Ory Kratos independent from the UI!
2. Kratos verifies the query parameters and responds with a redirect to a
   pre-configured URL such as `https://huddle.hsg.fs.tum.de/ui/profile`. A
   session cookie is set as well so that future requests by the user can be
   authenticated.
3. After the redirect, the web application UI takes over. For instance the user
   can set a new password on the profile site.

## Deployment and secret management

Let's first go through the deployment of Ory Kratos itself, which we do in
[Kubernetes](https://kubernetes.io/) with [Helm](https://helm.sh/) and
[Helmfile](https://github.com/roboll/helmfile).

Ory Kratos has a [Helm Chart](https://k8s.ory.sh/helm/kratos.html) that allows

> to install it in a cluster more or less conveniently. The interesting part here

Reviewer: ```suggestion
to install it in a cluster. The interesting part here
```

Sounds a bit like it wasn't convenient :)

                                email,
                                csrf_token
                            })
                            console.log(recoveryResponse.data);
                            authRequests.forEach(r => r.onFinished())
                            setAuthRequests([])
                        } catch (error) {
                            alertUnknownError(error)
                        }
                    }}>reset password</Button>
                <br />
                <Button onClick={() => {
                    authRequests.forEach(r => r.onFailed("User closed window"))
                    setAuthRequests([])
                }}>close</Button>
            </div>
        </div>
    )
}

export const logout = async () => {
    console.log('logout');
    const api = await getAPI()
    const resp = await api.createSelfServiceLogoutFlowUrlForBrowsers()
    console.log(resp.data);
    if (!resp) return
    const logoutResponse = await api.submitSelfServiceLogoutFlow(resp.data.logout_token)
    console.log(logoutResponse.data);
}
export const LogoutButton: React.FC = props => {
    return (
        <Button onClick={logout}>Logout</Button>
    )
}
...
```

> Is there a nicer way to handle the CSRF token? This feels a bit hacky...

Reviewer: ```suggestion
```

Doesn't seem relevant to the article?

  identity, ok := ctx.Value(identityKey).(*kc.Identity)
  if ok {
    traits, ok := identity.Traits.(map[string]interface{})
    if ok {
      return &Identity{Identity: *identity, traits: traits}, nil
    }
  }
  return nil, fmt.Errorf("no identity found in context")
}

func (i *Identity) GetTrait(key string) (string, bool) {
  val, ok := i.traits[key].(string)
  return val, ok
}
```

The traits property of the SDK identity type is untyped since it depends on the
identity schema, so in our own identity type we wrapped it in a map with a
custom getter for string fields.

In a resolver, we can access the identity like this:

```go
func (r *mutationResolver) SetMyDescription(ctx context.Context, description string) (bool, error) {
  me, err := auth.IdentityFromContext(ctx)
  if err != nil {
    return false, err
  }
  err = r.queries.SetDescription(context.Background(), sqlc.SetDescriptionParams{Description: description, ID: uuid.MustParse(me.Id)})
  return err == nil, err
}
```

Pretty elegant, right?

I hope to have given a nice introduction to Ory Kratos! Please
[feel free to comment in the discussion](https://github.com/ory/kratos/discussions/2356)

> on what we could do more elegantly! Also I'm interested in yoursolutions for

Reviewer: ```suggestion
on what we could do more elegantly! Also I'm interested in your solutions for
```

  secretsCipher: nasdlifuhlaiwd
  secretsCookie: lnaslicWERUOZSdicbl
  secretsDefault: ANSIULccbASKNXKAJ
  smtpConnectionURI: smtps://huddle.bot@zohomail.eu:qnIAsdHmsuU@smtp.zoho.eu:465
```

(of course I replaced the credentials by hitting the keyboard as wild :smiley:)

Now running `helmfile sync` will deploy Ory Kratos to our cluster!

Finally we need to publicly expose Ory Kratos using an ingress controller in
kubernetes:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: kratos-ingress
  annotations:
    kubernetes.io/ingress.class: 'nginx'
    nginx.ingress.kubernetes.io/rewrite-target: /$2
spec:
  tls:
    - hosts:
        - huddle.ridilla.eu
      secretName: huddle-ridilla-eu-prod-tls
    - hosts:
        - huddle.hsg.fs.tum.de
      secretName: huddle-hsg-fs-tum-de-prod-tls
  rules:
    - host: huddle.hsg.fs.tum.de
      http:
        paths:
          - path: /.ory/kratos(/|$)(.*)
            pathType: Prefix
            backend:
              service:
                name: kratos-public
                port:
                  number: 80
```

However two things feel a bit hacky about the deployment:

> 1. Even though we correctly set the base URL in the Ory Kratos config, we need

Reviewer 2: That's not a hacky configuration because you need to deal with URL rewrites in real-life configurations. As a solution, you can move Ory Kratos to a subdomain, like https://kratos.huddle... in that case, you'll proxy traffic from the route path of the subdomain to the root path of Kratos. On the other hand, you can leave it as is.

Reviewer 3: You are right, it's not that bad, I just think it's such a common issue that I would have expected the option to specify a base path in the kratos config, but I guess it's a matter of personal preferences.

   to do
   [regex magic to rewrite the path as if Ory Kratos was hosted under the root path](https://kubernetes.github.io/ingress-nginx/user-guide/ingress-path-matching/).

> 2. I was unable to find a way to put the Ory Kratos config in its own file

Reviewer 2: AFAIK `kratos.config` property is the correct way to pass kratos.config file using helm charts.

Reviewer 3: You are right again, it would only have been nice if someone knew some helm magic that allows to move the kratos config in its own file and template it to the kratos.config field. Like this there would be proper intellisense for the config fields.
